new/resources-anywhere.md

```diff
@@ -0,0 +1,211 @@
+
+### Accessing `File`
+
+Accessing top-level objects like `File` are one of the big reasons for this proposal.  Anything that extends from `Chef::Resource` *will* look up `Chef::Resource::X` before `::X` when a method types `X`.
+
+To fix this, we will move resources and providers in `Chef::Provider::*` and `Chef::Resource::*` into `Chef::Resources` (and use the inline syntax as well).  For addressing the obvious concerns, see the Errors and Deprecations section.
+
+### Removing `method_missing`
+
+Presently, when you type `my_resource` in the DSL, there is no actual method by that name; we intercept `method_missing`, perform a string transformation on the method name to get `MyResource`, and then try to get the `Chef::Resource::MyResource` class.  This causes confusing stack traces when the resource cannot be looked up, and is significantly more difficult to read.
+
+In this new world, a `Chef::Resources.my_resource` method is created, and `Chef::Resources` is mixed into the recipe DSL, so there is a real method.  `method_missing` never gets involved.  The methods generated look like this:
+
+```ruby
+def my_resource(name, created_at=nil, &block)
+  created_at ||= caller[0]
+  declare_resource(:my_resource, name, created_at, ::MyResource, &block)
+end
+```
+
+NOTE: This does not remove *all* of method_missing; in particular, we need to make definitions do something similar.  Add that to the proposal.
+
+### Declaring a Provider inline (Also Single-File Resources)
+
+In order to create a Resource, you must have a corresponding Provider.  The current algorithm takes `Chef::Resource::MyResource`, lops off `Chef::Resource::` entirely, and prepends `Chef::Provider::`.  This is The new algorithm will first look for a `Chef::Resource::MyResource::Provider` class, so that declaring the provider for a class will look like this for most people:
+
+```ruby
+class MyResource < Chef::Resource
+  ...
+
+  class Provider < Chef::Provider
+    ...
+  end
+end
+```
+
+This has several nice benefits:
+- Supports the goal of resources outside of `Chef::Resource`.
+- Encourages the creation of providers in the same file as the resource.
+- An absolute minimum of extra typing when linking a resource to a provider.
+
+### Creating a non-DSL resource
+
+Some users create base resources and .  This can be accomplished by calling `chef_recipe_dsl false` in the class, which will undef the DSL for it:
+
+```ruby
+class SuperAwesomeBaseResource < Chef::Resource
+  chef_recipe_dsl false
+
+  ...
+end
+```
+
+### Creating a DSL resource with a different name
+
+Sometimes the name of your class isn't the name that you need.  In that case, you can change the DSL with `chef_recipe_dsl`.
+
+```ruby
+class SuperMountPointResource < Chef::Resource
+  chef_recipe_dsl name: 'mount_point'
+end
+```
+
+This will remove the `super_mount_point_resource` DSL and add `mount_point` as DSL.
```

Comment:
I think you're colliding with provider_resolver and resource_resolver functionality that we already have here.

```ruby
class Chef
  class Resource
    class WindowsPackage < Chef::Resource::Package

      provides :package, os: "windows"
      provides :windows_package, os: "windows"
```